### PR TITLE
Spell scaling to item based spells.

### PR DESCRIPTION
**Pull request by GuybrushGit**
This PR handles all the item scaling related issues currently open.
All the fixes below are pretty straightforward except for the last 3. Would like to have some feedback concerning those.

Closes https://github.com/elysium-project/server/issues/426 https://github.com/elysium-project/server/issues/590 https://github.com/elysium-project/server/issues/593 https://github.com/elysium-project/server/issues/605 https://github.com/elysium-project/server/issues/614 https://github.com/elysium-project/server/issues/607 https://github.com/elysium-project/server/issues/608 https://github.com/elysium-project/server/issues/606 https://github.com/elysium-project/server/issues/719 https://github.com/elysium-project/server/issues/737 https://github.com/elysium-project/server/issues/846 https://github.com/elysium-project/server/issues/898

--------------------------

### Drain Life Melee Weapons
**Neretzek, The Blood Drinker / Ancient Hakkari Manslayer / Wraith Scythe / Fist of the Damned / Shadowstrike / Corrupted Ashbringer**

Multiple vanilla comments confirm 100% scaling including video evidence:
https://www.youtube.com/watch?v=Pk_kHHI0t4o&feature=youtu.be&t=90

>This weapon is SUPERB, because it scales 200% with spelldamage! (meaning with 500 spell power it absorbs 641 to 663)
>http://www.wowhead.com/item=21856/neretzek-the-blood-drinker#comments:id=7917:reply=43183

>The proc is affected by +spell damage equipment, which in my case (with melee gear) makes it damage+heal for 82-88
>I have +120 spell power and +274 healing with my current gear (I'm a Paladin). With a quick glance at my combat log, I saw: Your Drain Life heals you for 168.
>With a Paladin in spell damage gear this procs for 440-480 for me. (+378 in melee SD gear) 
>https://web.archive.org/web/20070409235359/http://www.thottbot.com/i19852

>It does also seem to be affected by +spelldamage gear - I have +96 and it usually heals me for about 145-150
>Whats even more unusual is it seems to have a 100% coefficient 
>https://web.archive.org/web/20060523231234/http://wow.allakhazam.com/item.html?witem=19852

>I have the entire blue dragonscale sett, and when this thing procs I get about 75-68 hp stolen/healed.
>http://web.archive.org/web/20070314054356/http://www.thottbot.com:80/i11920

>had randomly +7 dmg gear, then the life steal procced for 37 each hit instead of 30. removed the +7dmg, now it only procs for 30. 
>http://web.archive.org/web/20071029140830/http://thottbot.com:80/i10804

**Changes:** 100% coefficient given to all of these weapons.
Also fixed an issue with the direct health leech effect: it was receiving bonus healing from +healing gear.
The only player spell affected by this is Death Coil. 
For example a warlock with +500 "damage and healing" is currently getting +1000 bonus to the healing effect.

--------------------------

### Skullflame Shield

Multiple vanilla comments claim 100%:
>this shield takes 100% from +damage!  So with +60 in damage gear, you would be stealing about 100 life 
>+Spell damage adds to the life drain proc at a 1-to-1 ratio. With my 5/8 judgement and some other +spell damage items, the drain life is at about 255.
>I have +252 spell damage with my shaman and it do 287 damage 
>http://web.archive.org/web/20070408194508/http://www.thottbot.com:80/i1168

**Changes:** given 100% coefficient 

--------------------------

### Storm Gauntlets

Video evidence confirms 0% scaling.
https://www.youtube.com/watch?v=tDi_g4XDwBU

**Changes:** no action required

--------------------------

### Shard of the Fallen Star

Video evidence confirms 0% scaling.
https://www.youtube.com/watch?v=ay7RU0N-44Y
https://www.youtube.com/watch?v=Pk_kHHI0t4o&t=41

**Changes:** given 0% coefficient

--------------------------

### Zulian Slicer

Multiple vanilla comments claim it has 100% coefficient and ability to crit:
>just got this on my paladin, and im using 5/8 judgement, it seems to gain 100% bonus from damage/healing because it hits for around 230-250 and crits for 450ish
>http://web.archive.org/web/20070907094925/http://wow.allakhazam.com/item.html?witem=19901

>Being a mage and having over +400 dmg, it was kinda obvious (proc hits for 70-90, i was hitting for 475)
>http://www.disciplesguild.com/forums/topic/weapon-proc-pwnage/

**Changes:** given 100% coefficient and ability to crit

--------------------------

### Darkmoon Card: Heroism

Multiple vanilla comments claim it can crit:
>DOES critheal
>http://web.archive.org/web/20071211130436/http://wow.allakhazam.com:80/db/item.html?witem=19287;source=live

>The healing effect also has about 8% chance to crit
>http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Heroism?oldid=292291

**Changes:** given ability to crit, scaling remains at 0%

--------------------------

### Consumables
**Demonic Rune / Dark Rune / Whipper Root Tuber / Night Dragon's Breath**

Multiple comments and video evidence confirm they can crit:
https://www.youtube.com/watch?v=ScE8shXpz4Y&feature=youtu.be&t=253

>they can crit you, so be careful!
>while criting on the damage it does to me.
>http://web.archive.org/web/20061025231713/http://wow.allakhazam.com/db/item.html?witem=12662

>The damage effect can crit.
>http://db.vanillagaming.org/?item=20520#wh-comments

>Yes, they can crit
>http://www.lurkerlounge.com/forums/printthread.php?tid=4252

>The heal from this item can crit.
>http://db.vanillagaming.org/?item=11952#wh-comments

Nost added a rule to all potions and healthstones making them crit at a fixed 10% chance, it seems appropriate to include these items in that rule as well.

**Changes:** given ability to crit at a 10% rate

--------------------------

### Sulfuras, Hand of Ragnaros

Video evidence confirms no scaling.
Details here: https://github.com/elysium-project/server/issues/737

**Changes:** given 0% coefficient

--------------------------

### Ranged Weapons
**Hurricane / Bow of Searing Arrows / Heartseeking Crossbow / Dark Iron Rifle / Galganns Fireblaster / Quillshooter / Shell Launcher Shotgun / Venomstrike / Verdant Keepers Aim / Dwarven Hand Cannon**

One single comment from vanilla claims Hurricane has no scaling. 
One other comment claims firepower increases DHC damage.
Found no evidence for the rest of them, need someone with a retail account to check. 

>All procs ranged from 31-45 damage, so it appears that weapon procs do not benefit from bonuses to magical effects.
>http://web.archive.org/web/20061230212726/http://wow.allakhazam.com:80/db/item.html?witem=2824

>With a fire damage potion the Flaming Cannonball can crit for 100!
>http://www.wowhead.com/item=2099/dwarven-hand-cannon

**Changes:** given 0% coefficient to Hurricane, all others remain unchanged at 42%.

--------------------------

### Paladin T2 Judgement

Video evidence of no scaling and 2x crit damage
https://www.youtube.com/watch?v=ay7RU0N-44Y&feature=youtu.be&t=137

**Changes:** given 0% coefficient and 2x crit damage

--------------------------

### Demon Forged Breastplate

Current coefficient: 13%.
No relevant comments from vanilla, conflicting comments from TBC:

>03/22/2007 - My paladin with 300 spell damage ends up draining around 270 life from a target when this procs (50% coeff)
>http://db.vanillagaming.org/?item=12628#wh-comments

>Nov 13th 2007 - FACT: The drain life HOT is NOT affected by +dam/+heal.
>http://web.archive.org/web/20071130100327/http://wow.allakhazam.com:80/db/item.html?witem=12628

**Changes:** given 50% coefficient (12.5% per tick) since it seems unlikely that this was added in 2.0, but its debatable. 
**Needs staff decision**.

--------------------------

### Dragonbreath Chili

Current coefficient: 6%.
No comments from vanilla, a few from TBC: 

>Though I understand the chili does use your +spell damage or did for a bit
>http://www.lurkerlounge.com/forums/thread-3212.html?highlight=dragonbreath

>100% of your Fire +Spelldmg is added to the base dmg this proc does
>https://web.archive.org/web/20070314054427/http://www.thottbot.com:80/i12217

>Dragonbreath Chili no longer recieves a bonus from your +dmg/healing as of patch 2.2.0 
>http://www.wowhead.com/item=12217/dragonbreath-chili#comments

**Changes:** given 100% coefficient. Could this cause any imbalance? The proc rate is pretty low (5%). 
**Needs staff decision**.

--------------------------

### Darkmoon Card Maelstrom

Current coefficient: 42%.
Conflicting sources, some claim that it scales but don't say how much and the hits over 300 could be explained by the player using Stormstrike:

>i lied i dont think +spell dmg and effects works for maelstrom for some reason i thought it did
>http://web.archive.org/web/20070331193058/http://www.thottbot.com:80/i22321

>spell damage doesn't increase damage delt (2008/02/06)
>http://www.wowhead.com/item=19289/darkmoon-card-maelstrom#comments

>No, this was nerfed like 1 year ago
>http://web.archive.org/web/20070514023850/http://wow.allakhazam.com:80/db/item.html?witem=19289

>may be effected by +spell damage(hits over 300 have occurred on non-vunerable targets)
>http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=prev&oldid=368615

**Changes:** given 0% coefficient.
**Needs staff decision**.

### CODE REVIEW

**Comment by theloras** [comment]
Quick question regarding: 
_Paladin T2 Judgement

Video evidence of no scaling and 2x crit damage
https://www.youtube.com/watch?v=ay7RU0N-44Y&feature=youtu.be&t=137

Changes: given 0% coefficient and 2x crit damage_

I asked this on the dev discord but didn't get a straight answer - is the 8 piece set bonus currently able to also trigger weapon/item/trinket procs (ie: Nightfall spell vulverability debuff) like regular Judgement ability or will it be enabled to do so when these fixes go live?

**Comment by omniusaspirer** [comment]
I did extensive research into the ranged weapon proc scaling a while back, but had difficulty finding any evidence in either direction as you did. However, in contrast to the hurricane comment you found there is direct claims of at least one of these procs scaling with spellpower- the Dwarven Handcannon. 

http://www.wowhead.com/item=2099/dwarven-hand-cannon#comments

There's also *substantial* evidence that the procs should be critting with spell crit, which I haven't tested yet whether that's working correctly or not. 

In addition, proc rates are WAY off in the weapons I've tested. Galgann's fireblaster is one I can remember right now which should be proccing far more than the (15%ish if I remember right) that it is currently based on every comment I could find. 

What is the state of these weapons on retail WoW? Comments I found on other items indicated that these item based spells had remained largely untouched and those that scaled with spellpower were still doing so (see comments for Arcane Infused Gem for example). Assuming spellpower exists in the game still, retail will be the best resource for actually getting accurate data for these weapons in my opinion. I don't have an account, so someone who does would need to take the 20 minutes and test things out.

**Comment by omniusaspirer** [comment]
Also, the 10% crit rate for potions has absolutely zero evidence supporting it that I can find anywhere, which leads me to assume the percent was just a hack they applied since they knew that potions should crit but couldn't find solid data. 

In my own research, I came across a bug report on the Corecraft tracker which contained some discussion on this. They seemed to have concluded based on the spell mask that potion crit rate scaled off of physical crit. 

https://github.com/ccshiro/cc-buglist/issues/2581

**Comment by GuybrushGit** [comment]
Weird how the Hand Cannon would crit for 100 with 40 sp, that would mean 75% coeff which makes no sense so that value is probably wrong. Its still evidence that it scaled so it makes me unsure about the other ranged weapons, leaving them unchanged until someone with a retail account cares enough to check.

About the potion crit rate, that discussion on the cc bugtracker is only trying to figure out how it works on their server, its irrelevant to us.
In this discussion from 2008 it seems like the consensus is that it scales with spell crit, but theres also mention of the 10% fixed rate:
http://www.mmo-champion.com/threads/606235-Potions-Healthstones-and-other-restoring-items-Critical-age

The fixed rate does seem like the best solution so both physical and casters have a chance to crit but this discussion does not belong in this issue. These consumables should follow the potion crit rules and if in the future this rule changes they all change together.

**Comment by Mostprovidedlinks** [comment]
http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=prev&oldid=339109

and prior says it didn't scale with spell damage.

http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=next&oldid=321610
and later says it does scale.

The date you have shown says the latest at October 2006. The dates that say it does say as of November 12 2006. 

Conclusion: Darkmoon maelstrom should scale

**Comment by Mostprovidedlinks** [comment]
Zulian Slicer is stated to crit by x 1.9. Not x1.5

that should be mentioned too

**Comment by Mostprovidedlinks** [comment]
Demonic Rune / Dark Rune scale at player spell damage
it makes sense for them to be affected by player spell crit therefore they shouldn't have a fixed proc chance and should proc at the players spell crit

**Comment by omniusaspirer** [comment]
Did the original WoW client truncate or round up? If it rounded up, that would explain the oddly specific "hits for 100!" on the DHC and also reveal our correct coefficient for that item at least. 

49 (top end damage) + (40*.42857) = 66.1428. 

If that then crit, we'd be at 99.2142. Rounded up, we arrive at 100. 

Based on this possible explanation, I think it's reasonable to assume (before explicitly testing on retail) that this particular item should remain at the instant cast spell damage coefficient.

**Comment by omniusaspirer** [comment]
Another HUGE aspect of these weapons that's clarified a few times is that the procs trigger off of more than autoattacks. They should trigger off of ALL ranged abilities. Serpent Sting is included in this, and explicitly mentioned as having a chance to proc on each tick in the link below. 

http://www.wowhead.com/item=9412/galganns-fireblaster#comments

**Comment by thepparker** [comment]
All spells and abilities had a chance to round up depending on the mantissa

**Comment by GuybrushGit** [comment]
Can the @Mostprovidedlinks  comment spam be deleted? It offers no new information and just clutters up this issue.

**Comment by theloras** [comment]
@GuybrushGit - could you please answer my question from earlier regarding Tier 2 Judgement 8/8 set bonus.

With it being switched over to melee hit/crit table with your change here, will the set bonus' extra damage attack on all Judgements be able to trigger weapon procs such as Nightfall's spell vulnerability?

**Comment by Mostprovidedlinks** [comment]
Darkmoon Maelstrom should not be changed. The website you linked shows that in the revision it was made to scale with spell damage.

Your latest date shows
http://web.archive.org/web/20070514023850/http://wow.allakhazam.com:80/db/item.html?witem=19289 
Oct 29th 2006

http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=next&oldid=321610
 November 24 2006 and later says it does scale.

**Comment by Mostprovidedlinks** [comment]
Nightfalls proc should proc off 8/8 judgement since its a spell like the original judgement

**Comment by theloras** [comment]
I cant get an answer one way or the other...

**Comment by Itembugs** [comment]
silent fangs proc is also too low

**Comment by ImperfectCell** [comment]
the items still remain unfixed

**Comment by supasteed** [comment]
GuybrushGit can you add, storm gauntlets doesnt proc vengeance too

File: sql/migrations/20170604041201_world.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (26693, 1, 'Neretzek - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (24585, 1, 'Ancient Hakkari Manslayer - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (16414, 1, 'Wraith Scythe - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (18084, 1, 'Fist of the Damned - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (21170, 1, 'Shadowstrike - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29155, 1, 'Corrupted Ashbringer - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (18817, 1, 'Skullflame Shield - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (23682, 0, 'Heroism - No coefficient');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (23687, 0, 'Maelstrom - No coefficient');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: maelstrom should have 43% coefficient like it states in the november revision. 

http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=prev&oldid=368615 Scales in November 24, 2006 Revision as of 08:33

http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=prev&oldid=339113 The Revision of November 24, 2006 08:31 and beyond says it did.

http://wowwiki.wikia.com/wiki/Darkmoon_Card:_Maelstrom?direction=prev&oldid=339109
and before say it didn't

You are not correct saying it has a 0% coefficient. Here is more evidence with later dates showing this.

File: sql/migrations/20170604041201_world.sql [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (26789, 0, 'Shard of the Fallen Star - No coefficient');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (24251, 1, 'Zulian Slicer - 100% coefficient');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: this is correct

File: sql/migrations/20170604041201_world.sql [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (21162, 0, 'Sulfuras - No coefficient');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: this is correct

File: sql/migrations/20170604041201_world.sql [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (23590, 0, 'Paladin T2 Judgement - No coefficient');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: this is correct

File: sql/migrations/20170604041201_world.sql [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_mod (Id, DmgClass, Comment) VALUES (24251, 1, 'Zulian Slice: Magic DmgClass');
+INSERT INTO spell_mod (Id, DmgClass, Comment) VALUES (23590, 2, 'Paladin T2 Judgement: Melee DmgClass');
+
+DELETE FROM spell_bonus_data WHERE entry IN (26693,24585,16414,18084,21170,29155,18817,23682,23687,26789,24251,21162,23590,29502,29638,29640,29641,29644,29646,29647,29653,29655,15851,16608);
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (26693, 1, 'Neretzek - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (24585, 1, 'Ancient Hakkari Manslayer - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (16414, 1, 'Wraith Scythe - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (18084, 1, 'Fist of the Damned - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (21170, 1, 'Shadowstrike - Drain Life');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29155, 1, 'Corrupted Ashbringer - Drain Life');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: assuming drain life is a mirror version of the others which database says it is, yes you are right

File: sql/migrations/20170604041201_world.sql [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29638, 0, 'Bow of Searing Arrows - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29640, 0, 'Heartseeking Crossbow - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29641, 0, 'Dark Iron Rifle - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29644, 0, 'Galganns Fireblaster - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29646, 0, 'Quillshooter - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29647, 0, 'Shell Launcher Shotgun - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29653, 0, 'Venomstrike - No coefficient');
+-- INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (29655, 0, 'Verdant Keepers Aim - No coefficient');
+INSERT INTO spell_bonus_data (entry, direct_bonus, comments) VALUES (15851, 1, 'Dragonbreath Chili - 100% coefficient');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: this is correct

File: sql/migrations/20170604041201_world.sql [hunk 7]
```diff
@@ -0,0 +1,36 @@
+INSERT INTO migrations VALUES ('20170604041201');
+
+DELETE FROM spell_mod WHERE Id IN (16666,27869,15700,15701,23682,24251,23590);
+INSERT INTO spell_mod (Id, SpellFamilyName, Comment) VALUES (16666, 13, 'Demonic Rune: Potion Family so it can crit');
+INSERT INTO spell_mod (Id, SpellFamilyName, Comment) VALUES (27869, 13, 'Dark Rune: Potion Family so it can crit');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: This is wrong, dark rune is affected by player stats just like healing potion and demonic rune which is affected by player spell damage http://www.lurkerlounge.com/forums/printthread.php?tid=4252

Potions are stated to be affected by melee crit % 
#1177
ccshiro/cc-buglist#2581

Therefore if the player as 25% melee critical strike on him, the potion or rune will have a 25% chance

File: sql/migrations/20170604041201_world.sql [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_mod (Id, SpellFamilyName, Comment) VALUES (15700, 13, 'Whipper Root Tuber: Potion Family so it can crit');
+INSERT INTO spell_mod (Id, SpellFamilyName, Comment) VALUES (15701, 13, 'Night Dragons Breath: Potion Family so it can crit');
+INSERT INTO spell_mod (Id, DmgClass, Comment) VALUES (23682, 1, 'Heroism: Magic DmgClass so it can crit');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: This is correct

File: sql/migrations/20170604041201_world.sql [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+INSERT INTO spell_mod (Id, DmgClass, Comment) VALUES (24251, 1, 'Zulian Slice: Magic DmgClass');
+INSERT INTO spell_mod (Id, DmgClass, Comment) VALUES (23590, 2, 'Paladin T2 Judgement: Melee DmgClass');
```
Comment (supasteed) [inline sql/migrations/20170604041201_world.sql]: This is correct

**Comment by Mostprovidedlinks** [comment]
Can elysium please make sure these are all fixed in 1.10 ty

**Comment by vindruvor** [comment]
Do you have any info regarding Misplaced Servo Arm scaling?

**Pull request merged**